### PR TITLE
Dcr Test Functionality

### PR DESCRIPTION
**Pull request by t-etfali**
Matching current test extension in terms of functionality

### CODE REVIEW

File: main/logging.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	_, filePath, line, _ := runtime.Caller(2)
+
+	//Gets the line number and caller of the log message to identify what line logged
+	shortFileArr := strings.Split(filePath, "/")
+	shortFile := shortFileArr[len(shortFileArr)-1]
+
+	return fmt.Sprintf("[%s] [%s] [%s:%d] [%s]: %s", currentTime, version, shortFile, line, messageType, message)
+}
+
+type customOperationLogger struct {
```
Comment (larohra) [inline main/logging.go]: try re-using the same logger for this too (if not too much trouble). Dont spend too much time on it though

File: main/logging.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+func formatOperationMessage(message string) string {
+	currentTime := time.Now().UTC().Format("2006-01-02T15:04:05.999Z")
+	return fmt.Sprintf("[%s] [%s]: %s", currentTime, version, message)
+}
+
+func initAllLogging() (generalFile *os.File, operationFile *os.File, generalErr error, operationErr error) {
+	generalFile, generalErr = initGeneralLogging()
+	if generalErr != nil {
+		fmt.Printf("Error opening the general logfile. %+v", generalErr)
+		generalErr = errors.Wrap(generalErr, "Failed to open general logfile")
```
Comment (larohra) [inline main/logging.go]: Any particular reason for continuing with `initOperationLogging` even if `initGeneralLogging` fails? Why not just fail fast and proceed?

File: main/logging.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		fmt.Printf("Error opening the operation logfile. %+v", operationErr)
+		operationErr = errors.Wrap(operationErr, "Failed to open operation logfile")
+	}
+	return generalFile, operationFile, generalErr, operationErr
+}
+
+func initGeneralLogging() (*os.File, error) {
+	handlerEnv, handlerEnvErr := settings.GetHandlerEnvironment()
+
+	logfileLogName := extensionName + "-" + version + ".log"
```
Comment (larohra) [inline main/logging.go]: The General logger should be version-independent, all versions should log to the same file

File: main/logging.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+	//Caller(1) is the customlogger Print function in this file
+	//Caller(2) is the line that called the logger's print function
+	_, filePath, line, _ := runtime.Caller(2)
+
+	//Gets the line number and caller of the log message to identify what line logged
+	shortFileArr := strings.Split(filePath, "/")
+	shortFile := shortFileArr[len(shortFileArr)-1]
+
+	return fmt.Sprintf("[%s] [%s] [%s:%d] [%s]: %s", currentTime, version, shortFile, line, messageType, message)
```
Comment (larohra) [inline main/logging.go]: Does golang automatically pick up variables from other modules without importing? Like I dont see any imports for version?

Comment (larohra) [reply main/logging.go]: Also according to the [naming convention ](https://stackoverflow.com/questions/22688906/go-naming-conventions-for-const), I think you should rename version to `Version`

File: main/logging.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+func (l *customOperationLogger) Printf(format string, v ...interface{}) {
+	l.logger.Printf(formatOperationMessage(fmt.Sprintf(format, v...)))
+}
+
+func formatOperationMessage(message string) string {
+	currentTime := time.Now().UTC().Format("2006-01-02T15:04:05.999Z")
+	return fmt.Sprintf("[%s] [%s]: %s", currentTime, version, message)
+}
+
+func initAllLogging() (generalFile *os.File, operationFile *os.File, generalErr error, operationErr error) {
```
Comment (larohra) [inline main/logging.go]: If something is being exported then it should start with a capital letter, else start with lower case.

File: main/main.go [hunk 6]
```diff
@@ -39,159 +45,204 @@ const (
...
 }
 
-func uninstall() {
-	infoLogger.Println("Uninstalled Successfully.")
+// extension specific ProtectedSettings
+type extensionPrivateSettings struct {
+	SecretString       string   `json:"secretString"`
+	SecretScript       string   `json:"secretScript"`
+	FileURLs           []string `json:"fileUris"`
+	StorageAccountName string   `json:"storageAccountName"`
```
Comment (larohra) [inline main/main.go]: You dont need most of this extra stuff, just keep `SecretString` and remove the rest

File: main/main.go [hunk 7]
```diff
@@ -39,159 +45,204 @@ const (
...
+	StorageAccountKey  string   `json:"storageAccountKey"`
 }
 
-func update() {
-	infoLogger.Println("Updated Successfully.")
-}
+func install() {
+	operation := "install"
+	infoLogger.Printf("Extension MrSeq: %d, Environment MrSeq: %d", extensionMrSeq, environmentMrSeq)
+	operationLogger.Printf("[Seq Num: %d] [operation: %s]", environmentMrSeq, operation)
```
Comment (larohra) [inline main/main.go]: Set both the Operation and SeqNo as format in the logging module and just pass this data as params instead to avoid confusion and to ensure we always log those values

File: main/main.go [hunk 8]
```diff
@@ -39,159 +45,204 @@ const (
...
 
-func parseJSON(filename string) error {
-	//	Open the provided file
-	jsonFile, err := os.Open(filename)
+	err := status.ReportTransitioning(environmentMrSeq, operation, "installation in progress")
+	infoLogger.Println("Installation in progress")
 	if err != nil {
-		return errors.Wrapf(err, "Failed to open \"%s\"", filename)
+		errorLogger.Printf("Status reporting error: %+v", err)
+		os.Exit(statusReportingError)
```
Comment (larohra) [inline main/main.go]: Have a central error logging mechanism instead of having a separate `Exit` for each function. In that central error logger, make sure to `status.ReportFailure` properly

File: main/main.go [hunk 9]
```diff
@@ -39,159 +45,204 @@ const (
...
-
-		infoLogger.Println(key, reverseValue)
+	err = status.ReportSuccess(environmentMrSeq, operation, "installation is complete")
+	if err != nil {
+		errorLogger.Printf("Status reporting error: %+v", err)
+		os.Exit(statusReportingError)
 	}
-	return nil
+	infoLogger.Println("Installation is complete")
+	os.Exit(successfulExecution)
```
Comment (larohra) [inline main/main.go]: Similarly, dont exit separately from each function even on success, only exit from the main() function so that its common for all functions

File: main/main.go [hunk 10]
```diff
@@ -39,159 +45,204 @@ const (
...
-type ProtectedSettings struct {
-	SecretString       string   `json:"secretString"`
-	SecretScript       string   `json:"secretScript"`
-	FileURLs           []string `json:"fileUris"`
-	StorageAccountName string   `json:"storageAccountName"`
-	StorageAccountKey  string   `json:"storageAccountKey"`
-}
+func disable() {
+	operation := "disable"
+	infoLogger.Printf("Extension MrSeq: %d, Environment MrSeq: %d", extensionMrSeq, environmentMrSeq)
```
Comment (larohra) [inline main/main.go]: Most of the code for all these operations are similar (except for maybe enable). Try to maybe create a common utility function for all instead of having so much code duplication to make it easier to read and manage.

File: test.json [hunk 11]
```diff
@@ -1,15 +1,3 @@
...
-        "key4" : "string1",
-        "key5" : "Is String 1 above?",
-        "key6" : "string 1 question above?",
-        "key7" : "racecar"
-    },
-    "notKeys": {
-        "key1" : "foo",
-        "key2" : "bar"
-    }
+    "failCommands": ["enable"]
```
Comment (larohra) [inline test.json]: Remove this file and add it to the .gitignore

**Review by larohra — CHANGES_REQUESTED** [review]

File: main/logging.go (hunk 4 above)
Comment (t-etfali) [reply main/logging.go]: When other modules are declared as part of the same package then it can pick up values without importing (since everything is in package main)

Comment (larohra) [reply main/logging.go]: Ahh interesting, so its part of the same package then. Then not sure if the naming convention applies here.

Comment (larohra) [reply main/logging.go]: You can leave the naming as is then, sorry for the confusion!

File: main/logging.go [hunk 12]
```diff
@@ -20,69 +20,68 @@ var (
...
+	customLogger
 }
 
-func (l *customGeneralLogger) Println(v ...interface{}) {
-	l.logger.Println(formatLoggingMessage(l.operation, fmt.Sprint(v...)))
+type customOperationLogger struct {
+	customLogger
 }
 
+func (l *customGeneralLogger) Print(v ...interface{}) {
```
Comment (larohra) [inline main/logging.go]: A suggestion, instead of creating 2 new types (customGeneralLogger  and customOperationLogger), wouldn't it be easier to add another variable to the customLogger type? Maybe something like - 

```go
type customLogger struct {
	logger    *log.Logger		
	operation string
        mode enum // Can either be Operations or General
}
```

This way you wouldn't have to duplicate the Println, Printf, Print functions for both the loggers.

File: main/logging.go [hunk 13]
```diff
@@ -20,69 +20,68 @@ var (
...
-	l.logger.Printf(formatOperationMessage(fmt.Sprintf(format, v...)))
+	l.logger.Printf(formatLoggingMessage(*l, l.operation, fmt.Sprintf(format, v...)))
 }
 
-func formatOperationMessage(message string) string {
-	currentTime := time.Now().UTC().Format("2006-01-02T15:04:05.999Z")
-	return fmt.Sprintf("[%s] [%s]: %s", currentTime, version, message)
-}
+func formatLoggingMessage(l interface{}, messageType string, message string) string {
+	switch loggerType := l.(type) {
```
Comment (larohra) [inline main/logging.go]: If you add a mode as above, you can change the switch to be on the mode instead

File: main/logging.go [hunk 14]
```diff
@@ -20,69 +20,68 @@ var (
...
+	case customGeneralLogger:
+		currentTime := time.Now().UTC().Format("2006-01-02T15:04:05.999Z")
```
Comment (larohra) [inline main/logging.go]: Move the duplicate fields outside the switch scope (like the currentTime) to avoid code duplication as much as possible

File: main/logging.go [hunk 15]
```diff
@@ -20,69 +20,68 @@ var (
...
-	return generalFile, operationFile, generalErr, operationErr
 }
 
-func initGeneralLogging() (*os.File, error) {
+func initGeneralLogging() (file *os.File, err error) {
 	handlerEnv, handlerEnvErr := settings.GetHandlerEnvironment()
 
-	logfileLogName := extensionName + "-" + version + ".log"
+	logfileLogName := extensionName + ".log"
 	if handlerEnvErr != nil {
```
Comment (larohra) [inline main/logging.go:74]: There is code duplication from line 85-92 and 114-121. Can you make a small helper function out of it instead?

File: main/logging.go [hunk 16]
```diff
@@ -20,69 +20,68 @@ var (
...
 
-	operationFile, operationErr = initOperationLogging()
-	if operationErr != nil {
-		fmt.Printf("Error opening the operation logfile. %+v", operationErr)
-		operationErr = errors.Wrap(operationErr, "Failed to open operation logfile")
+		return fmt.Sprintf("[%s] [%s] [%s:%d] [%s]: %s", currentTime, version, shortFile, line, messageType, message)
+
+	case customOperationLogger:
+		currentTime := time.Now().UTC().Format("2006-01-02T15:04:05.999Z")
+		return fmt.Sprintf("[%s] [%s]: [Seq Num: %d] [operation: %s]", currentTime, version, environmentMrSeq, message)
```
Comment (larohra) [inline main/logging.go]: Add operation in this too

File: main/logging.go [hunk 17]
```diff
@@ -20,69 +20,68 @@ var (
...
 func (l *customOperationLogger) Printf(format string, v ...interface{}) {
-	l.logger.Printf(formatOperationMessage(fmt.Sprintf(format, v...)))
+	l.logger.Printf(formatLoggingMessage(*l, l.operation, fmt.Sprintf(format, v...)))
 }
 
-func formatOperationMessage(message string) string {
-	currentTime := time.Now().UTC().Format("2006-01-02T15:04:05.999Z")
-	return fmt.Sprintf("[%s] [%s]: %s", currentTime, version, message)
-}
+func formatLoggingMessage(l interface{}, messageType string, message string) string {
```
Comment (larohra) [inline main/logging.go]: NIT: Rename messageType to Operation or Operation to messageType in the definition of customLogger

File: main/logging.go [hunk 18]
```diff
@@ -121,16 +120,34 @@ func initOperationLogging() (*os.File, error) {
...
-	file, err := os.OpenFile(operationLogfile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	file, err = os.OpenFile(operationLogfile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
 	if err != nil {
 		return nil, errors.Wrapf(err, "Failed to create/open %s", operationLogfile)
 	}
 
-	//Sample: [2020-08-18T20:29:16.079902Z] [1.0.0.0] [main.go:148] [INFO]: Test1
-	operationLogger = customOperationLogger{log.New(file, "", 0)}
+	//Sample: [2020-08-20T23:29:30.676Z] [1.0.0.2]: [Seq Num: 0] [operation: install]
+	operationLogger = customOperationLogger{customLogger{log.New(file, "", 0), ""}}
```
Comment (larohra) [inline main/logging.go]: Add operation to this log too, make it just info by default

File: main/main.go [hunk 19]
```diff
@@ -56,162 +52,125 @@ type extensionPublicSettings struct {
...
-	StorageAccountName string   `json:"storageAccountName"`
-	StorageAccountKey  string   `json:"storageAccountKey"`
+	SecretString string `json:"secretString"`
 }
 
-func install() {
-	operation := "install"
+func reportStatus(statusType string, operation string, message string) {
+	switch statusType {
+	case "success":
```
Comment (larohra) [inline main/main.go]: NIT: Create an enum of statuses (success, failure, transitioning)

File: main/main.go [hunk 20]
```diff
@@ -56,162 +52,125 @@ type extensionPublicSettings struct {
...
 
-	err = settings.GetExtensionSettings(environmentMrSeq, &publicSettings, &protectedSettings)
+	err := settings.GetExtensionSettings(environmentMrSeq, &publicSettings, &protectedSettings)
 	if err != nil {
-		status.ReportError(environmentMrSeq, operation, err.Error())
-		errorLogger.Printf("%+v", err)
-		os.Exit(settingsNotFoundError)
+		errorMessage := fmt.Sprintf("Error getting settings: %+v", err)
+		errorLogger.Println(errorMessage)
+		executionErrors = append(executionErrors, errorMessage)
```
Comment (larohra) [inline main/main.go:145]: Shouldn't you return here if there's an error?

File: main/main.go [hunk 21]
```diff
@@ -56,162 +52,125 @@ type extensionPublicSettings struct {
...
+
+	//	Unmarshall the bytes from the JSON file
+	byteValue, _ := ioutil.ReadAll(jsonFile)
+	var jsonData map[string][]string
+	json.Unmarshal([]byte(byteValue), &jsonData)
+
+	return nil
+}
+
+func reportExecutionStatus() {
```
Comment (larohra) [inline main/main.go:117]: This function is never called

File: main/main.go [hunk 22]
```diff
@@ -56,162 +52,125 @@ type extensionPublicSettings struct {
...
+		warningLogger.Println("Status report type not recognized")
+	}
+
+}
+
+func testCommand(operation string) {
 	infoLogger.Printf("Extension MrSeq: %d, Environment MrSeq: %d", extensionMrSeq, environmentMrSeq)
-	operationLogger.Printf("[Seq Num: %d] [operation: %s]", environmentMrSeq, operation)
+	operationLogger.Println(operation)
+	reportStatus("transitioning", operation, fmt.Sprintf("%s in progress", operation))
```
Comment (larohra) [inline main/main.go]: This line is not really needed

File: main/main.go [hunk 23]
```diff
@@ -17,15 +15,26 @@ import (
 )
 
 var (
-	version            = "1.0.0.0"
-	extensionShortName = "GATestExt"
+	versionMajor    = "1"
+	versionMinor    = "0"
+	versionBuild    = "0"
+	versionRevision = "2"
```
Comment (larohra) [inline main/main.go]: You shouldn't bump up version if there's no release. 1 release can have many features, we should send another PR just for version bump whenever we're releasing.
Also, this version should always point to the last released version which was 1.0.0.1 in our case

**Review by larohra — CHANGES_REQUESTED** [review]

File: main/logging.go [hunk 24]
```diff
@@ -66,72 +57,61 @@ func formatLoggingMessage(l interface{}, messageType string, message string) str
...
 
-		return fmt.Sprintf("[%s] [%s] [%s:%d] [%s]: %s", currentTime, version, shortFile, line, messageType, message)
-
-	case customOperationLogger:
-		currentTime := time.Now().UTC().Format("2006-01-02T15:04:05.999Z")
-		return fmt.Sprintf("[%s] [%s]: [Seq Num: %d] [operation: %s]", currentTime, version, environmentMrSeq, message)
-
+		return fmt.Sprintf("[%s] [%s] [%s:%d] [%s]: %s", currentTime, version, shortFile, line, messagePrefix, message)
+	case operationLoggerMode:
+		// Message type for the operation logger refers to one of either install,enable, etc
```
Comment (larohra) [inline main/logging.go]: NIT: Change `Message Type` to `Message Prefix` to match with the parameters

**Review by larohra — CHANGES_REQUESTED** [review]
Some very minor comments

File: main/logging.go (hunk 24 above)
Comment (larohra) [reply main/logging.go]: Ping on this

**Review by larohra — APPROVED** [review]
